src/charm.py

```diff
@@ -54,25 +55,27 @@ def __init__(self, *args):
         self.framework.observe(self.on[ZK].relation_changed, self._on_config_changed)
         self.framework.observe(self.on[ZK].relation_broken, self._on_zookeeper_broken)
 
-        self.framework.observe(self.on.set_password_action, self._set_password_action)
+        self.framework.observe(getattr(self.on, "set_password_action"), self._set_password_action)
 
     @property
-    def peer_relation(self) -> Relation:
+    def peer_relation(self) -> Optional[Relation]:
         """The cluster peer relation."""
         return self.model.get_relation(PEER)
 
     @property
-    def app_peer_data(self) -> Dict:
+    def app_peer_data(self) -> MutableMapping[str, str]:
```

Comment:
Why go too generic with `MutableMapping` instead of going with `Dict[str, str]`?

Reply:
I think you should be safe using a dict here, `MutableMapping` is just an abstract type

Reply:
You're right, but type-hint warnings clog up the screen.